File: android/app/src/main/java/com/videoconverter/android/domain/Queue.kt
package com.videoconverter.android.domain

fun splitImportable(
    sources: List<MediaInfo>,
    cannotTranscode: String = "Could not convert this file",
): Pair<List<MediaInfo>, List<SkippedSource>> {
    val accepted = mutableListOf<MediaInfo>()
    val skipped = mutableListOf<SkippedSource>()
    for (source in sources) {
        if (source.importable) {
            accepted += source
        } else {
            skipped += SkippedSource(
                sourceUri = source.sourceUri,
                displayName = source.displayName,
                reason = source.error ?: cannotTranscode,
            )
        }
    }
    return accepted to skipped
}

fun configForSource(config: OutputConfig, media: MediaInfo): OutputConfig =
    if (media.trimStartSecs != null || media.trimEndSecs != null) {
        config.copy(
            trimStartSecs = media.trimStartSecs,
            trimEndSecs = media.trimEndSecs,
        )
    } else {
        config
    }

fun enqueueJobs(
    sources: List<MediaInfo>,
    config: OutputConfig,
    outputDir: String,
    nextId: () -> String,
    exists: (String) -> Boolean,
    cannotTranscode: String = "Could not convert this file",
    selectOutput: String = "Choose an output folder first",
    validateCopy: ValidateCopy = ValidateCopy(),
    unknownPreset: (String) -> String = { "Unknown preset: $it" },
    clock: () -> String = ::outputCollisionStamp,
    nowMs: () -> Long = { System.currentTimeMillis() },
    concatNeedsTwo: String = "Add at least two videos",
    concatTooMany: String = "You can merge up to 20 videos",
    concatMissingVideo: String = "Each clip needs a video track",
): Result<EnqueueReport> = runCatching {
    if (outputDir.isBlank()) {
        throw IllegalArgumentException(selectOutput)
    }

    val resolved = resolveConfig(config, unknownPreset).getOrThrow()
    if (isVideoConcatPreset(config.preset)) {
        return@runCatching enqueueConcatJobs(
            sources = sources,
            config = config,
            resolved = resolved,
            outputDir = outputDir,
            nextId = nextId,
            exists = exists,
            cannotTranscode = cannotTranscode,
            validateCopy = validateCopy,
            clock = clock,
            nowMs = nowMs,
            concatNeedsTwo = concatNeedsTwo,
            concatTooMany = concatTooMany,
            concatMissingVideo = concatMissingVideo,
        )
    }
    val (accepted, initialSkipped) = splitImportable(sources, cannotTranscode)
    val skipped = initialSkipped.toMutableList()
    val jobs = mutableListOf<Job>()
    val allocated = mutableSetOf<String>()

    for (media in accepted) {
        val validation = validate(resolved, media, validateCopy)
        if (validation.isFailure) {
            skipped += SkippedSource(
                sourceUri = media.sourceUri,
                displayName = media.displayName,
                reason = validation.exceptionOrNull()?.message ?: cannotTranscode,
            )
            continue
        }

        val outputPath = allocateOutputPath(
            outputDir = outputDir,
            stem = sourceStem(media.displayName),
            ext = resolved.extension,
            exists = { candidate ->
                val partial = partialOutputPath(candidate)
                exists(candidate) ||
                    exists(partial) ||
                    candidate in allocated ||
                    partial in allocated
            },
            clock = clock,
        )
        allocated += outputPath
        allocated += partialOutputPath(outputPath)

        jobs += Job(
            id = nextId(),
            sourceUri = media.sourceUri,
            displayName = media.displayName,
            outputPath = outputPath,
            status = JobStatus.Queued,
            progress = 0.0,
            error = null,
            config = configForSource(config, media),
            media = media,
            createdAtEpochMs = nowMs(),
        )
    }

    EnqueueReport(jobs = jobs, skipped = skipped)
}

private fun enqueueConcatJobs(
    sources: List<MediaInfo>,
    config: OutputConfig,
    resolved: ResolvedConfig,
    outputDir: String,
    nextId: () -> String,
    exists: (String) -> Boolean,
    cannotTranscode: String,
    validateCopy: ValidateCopy,
    clock: () -> String,
    nowMs: () -> Long,
    concatNeedsTwo: String,
    concatTooMany: String,
    concatMissingVideo: String,
): EnqueueReport {
    val (accepted, skipped) = splitImportable(sources, cannotTranscode)
    if (skipped.isNotEmpty()) {
        return EnqueueReport(jobs = emptyList(), skipped = skipped)
    }
    if (accepted.size < 2) {
        return EnqueueReport(
            jobs = emptyList(),
            skipped = accepted.map {
                SkippedSource(it.sourceUri, it.displayName, concatNeedsTwo)
            },
        )
    }
    if (accepted.size > VIDEO_CONCAT_MAX_SOURCES) {
        return EnqueueReport(
            jobs = emptyList(),
            skipped = accepted.map {
                SkippedSource(it.sourceUri, it.displayName, concatTooMany)
            },
        )
    }
    val invalid = mutableListOf<SkippedSource>()
    for (media in accepted) {
        try {
            concatTarget(media, concatMissingVideo)
            if (media.videoCodec == null) {
                throw IllegalArgumentException(concatMissingVideo)
            }
            validate(resolved, media, validateCopy).getOrThrow()
        } catch (error: Exception) {
            invalid += SkippedSource(
                sourceUri = media.sourceUri,
                displayName = media.displayName,
                reason = error.message ?: cannotTranscode,
            )
        }
    }
    if (invalid.isNotEmpty()) {
        return EnqueueReport(jobs = emptyList(), skipped = invalid)
    }

    val outputPath = allocateOutputPath(
        outputDir = outputDir,
        stem = concatOutputStem(accepted[0].displayName),
        ext = resolved.extension,
        exists = { candidate ->
            val partial = partialOutputPath(candidate)
            exists(candidate) || exists(partial)
        },
        clock = clock,
    )
    val first = accepted[0]
    val job = Job(
        id = nextId(),
        sourceUri = first.sourceUri,
        displayName = first.displayName,
        outputPath = outputPath,
        status = JobStatus.Queued,
        progress = 0.0,
        error = null,
        config = config.copy(concatSourceUris = accepted.map { it.sourceUri }),
        media = first,
        createdAtEpochMs = nowMs(),
        concatMedias = accepted,
    )
    return EnqueueReport(jobs = listOf(job), skipped = emptyList())
}

fun enqueueDocumentJobs(
    sources: List<MediaInfo>,
    config: OutputConfig,
    outputDir: String,
    nextId: () -> String,
    exists: (String) -> Boolean,
    cannotTranscode: String = "Could not convert this file",
    selectOutput: String = "Choose an output folder first",
    cannotReadPages: String = "Could not read the page count",
    clock: () -> String = ::outputCollisionStamp,
    nowMs: () -> Long = { System.currentTimeMillis() },
): Result<EnqueueReport> = runCatching {
    if (outputDir.isBlank()) {
        throw IllegalArgumentException(selectOutput)
    }

    val (accepted, initialSkipped) = splitImportable(sources, cannotTranscode)
    val skipped = initialSkipped.toMutableList()
    val jobs = mutableListOf<Job>()
    val allocated = mutableSetOf<String>()
    val extension = documentExtension(config.preset, config.container)

    for (media in accepted) {
        if (needsPdfPageCount(config.preset) && media.pageCount == null) {
            skipped += SkippedSource(
                sourceUri = media.sourceUri,
                displayName = media.displayName,
                reason = cannotReadPages,
            )
            continue
        }

        val outputPath = allocateOutputPath(
            outputDir = outputDir,
            stem = sourceStem(media.displayName),
            ext = extension,
            exists = { candidate ->
                val partial = partialOutputPath(candidate)
                exists(candidate) ||
                    exists(partial) ||
                    candidate in allocated ||
                    partial in allocated
            },
            clock = clock,
        )
        allocated += outputPath
        allocated += partialOutputPath(outputPath)

        jobs += Job(
            id = nextId(),
            sourceUri = media.sourceUri,
            displayName = media.displayName,
            outputPath = outputPath,
            status = JobStatus.Queued,
            progress = 0.0,
            error = null,
            config = configForSource(config, media),
            media = media,
            createdAtEpochMs = nowMs(),
        )
    }

    EnqueueReport(jobs = jobs, skipped = skipped)
}

private fun needsPdfPageCount(preset: String): Boolean =
    preset == "pdf-image" ||
        preset == "pdf-txt" ||
        preset == "pdf-compress" ||
        preset == "pdf-split"

fun markInterrupted(
    jobs: List<Job>,
    interrupted: String = "Conversion was interrupted",
): List<Job> = jobs.map { job ->
    if (job.status == JobStatus.Running) {
        job.copy(status = JobStatus.Failed, error = interrupted)
    } else {
        job
    }
}

File: ios/LiteTrans/Domain/Queue.swift
import Foundation

public func splitImportable(_ sources: [MediaInfo], cannotTranscode: String = "Could not convert this file") -> ([MediaInfo], [SkippedSource]) {
    var accepted: [MediaInfo] = []
    var skipped: [SkippedSource] = []
    for source in sources {
        if source.importable {
            accepted.append(source)
        } else {
            skipped.append(.init(sourceUri: source.sourceUri, displayName: source.displayName, reason: source.error ?? cannotTranscode))
        }
    }
    return (accepted, skipped)
}

public func configForSource(_ config: OutputConfig, media: MediaInfo) -> OutputConfig {
    var next = config
    if allowsTrim(preset: config.preset), media.trimStartSecs != nil || media.trimEndSecs != nil {
        next.trimStartSecs = media.trimStartSecs
        next.trimEndSecs = media.trimEndSecs
    }
    return next
}

public func occupiedOutputPaths(_ jobs: [Job]) -> Set<String> {
    var paths = Set<String>()
    for job in jobs {
        for path in resolvedOutputPaths(job) {
            paths.insert(path)
            paths.insert(partialOutputPath(path))
        }
    }
    return paths
}

public func outputFileDeletionPaths(job: Job, remainingJobs: [Job]) -> [String] {
    guard job.status == .completed else { return [] }
    let occupied = occupiedOutputPaths(remainingJobs)
    var deletion: [String] = []
    for path in resolvedOutputPaths(job) where !occupied.contains(path) {
        deletion.append(path)
        deletion.append(partialOutputPath(path))
    }
    return deletion
}

private func resolvedOutputPaths(_ job: Job) -> [String] {
    job.outputPaths.isEmpty ? [job.outputPath].compactMap { $0 } : job.outputPaths
}

public func shouldDeleteImportedSource(
    sourceUri: String,
    remainingJobs: [Job],
    sessionSources: [MediaInfo]
) -> Bool {
    remainingJobs.allSatisfy { $0.sourceUri != sourceUri }
        && sessionSources.allSatisfy { $0.sourceUri != sourceUri }
}

public func enqueueJobs(
    sources: [MediaInfo],
    config: OutputConfig,
    outputDir: String,
    nextId: () -> String,
    exists: (String) -> Bool,
    existingJobs: [Job] = [],
    outputKind: OutputKind = .downloads,
    nowMs: () -> Int64 = { currentEpochMs() }
) throws -> EnqueueReport {
    if outputDir.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
        throw LiteTransError.blankOutputDir
    }
    let resolved = try resolveConfig(config)
    let (accepted, initialSkipped) = splitImportable(sources)
    var skipped = initialSkipped
    var jobs: [Job] = []
    var allocated = occupiedOutputPaths(existingJobs)
    for media in accepted {
        if documentSourceKind(media.displayName) == .word || documentSourceKind(media.displayName) == .excel {
            skipped.append(
                .init(
                    sourceUri: media.sourceUri,
                    displayName: media.displayName,
                    reason: LiteTransError.officeNotAvailable.localizedDescription
                )
            )
            continue
        }
        do {
            try validate(resolved, media: media)
        } catch {
            skipped.append(.init(sourceUri: media.sourceUri, displayName: media.displayName, reason: error.localizedDescription))
            continue
        }
        let isTaken: (String) -> Bool = { candidate in
            let partial = partialOutputPath(candidate)
            return exists(candidate) || exists(partial) || allocated.contains(candidate) || allocated.contains(partial)
        }
        let stem = sourceStem(media.displayName)
        let ext = resolved.preset == "image-compress"
            ? keptImageExtension(media.displayName)
            : resolved.extension
        let count = outputCount(preset: resolved.preset, media: media)
        var outputPaths: [String] = []
        if count > 1 {
            for index in 1...count {
                let numbered = numberedOutputName(stem: stem, index: index, ext: ext)
                let path = allocateOutputPath(
                    outputDir: outputDir,
                    stem: sourceStem(numbered),
                    ext: ext,
                    exists: isTaken
                )
                outputPaths.append(path)
                allocated.insert(path)
                allocated.insert(partialOutputPath(path))
            }
        } else {
            let path = allocateOutputPath(outputDir: outputDir, stem: stem, ext: ext, exists: isTaken)
            outputPaths.append(path)
            allocated.insert(path)
            allocated.insert(partialOutputPath(path))
        }
        jobs.append(Job(
            id: nextId(),
            sourceUri: media.sourceUri,
            displayName: media.displayName,
            outputPath: outputPaths[0],
            status: .queued,
            progress: 0,
            error: nil,
            config: configForSource(config, media: media),
            media: media,
            outputPaths: outputPaths,
            outputKind: outputKind,
            createdAtEpochMs: nowMs()
        ))
    }
    return EnqueueReport(jobs: jobs, skipped: skipped)
}

public func markInterrupted(_ jobs: [Job], interrupted: String = "Conversion was interrupted") -> [Job] {
    jobs.map { job in
        job.status == .running
            ? Job(
                id: job.id,
                sourceUri: job.sourceUri,
                displayName: job.displayName,
                outputPath: job.outputPath,
                status: .failed,
                progress: job.progress,
                error: interrupted,
                config: job.config,
                media: job.media,
                outputPaths: job.outputPaths,
                outputKind: job.outputKind,
                createdAtEpochMs: job.createdAtEpochMs
            )
            : job
    }
}
